core/src/main/java/io/grpc/ServiceProviders.java

```diff
@@ -0,0 +1,149 @@
+   * {@code androidClasses}.
+   * The list is sorted in descending priority order.
+   */
+  public static <T extends ServiceProvider> List<T> loadAll(
+      Class<T> klass, List<String> androidClasses, ClassLoader cl) {
+    Iterable<T> candidates;
+    if (isAndroid(cl)) {
+      candidates = getCandidatesViaHardCoded(klass, androidClasses, cl);
+    } else {
+      candidates = getCandidatesViaServiceLoader(klass, cl);
+    }
+    List<T> list = new ArrayList<T>();
+    Iterator<T> iter = candidates.iterator();
+    while (iter.hasNext()) {
+      try {
+        T current = iter.next();
+        if (!current.isAvailable()) {
+          continue;
+        }
+        list.add(current);
+      } catch (Throwable t) {
+        // The iterator from ServiceLoader may throw ServiceConfigurationError
+        logger.log(Level.SEVERE, "caught exception trying to load: " + klass, t);
+      }
+    }
+    // Sort descending based on priority.
+    Collections.sort(list, Collections.reverseOrder(new Comparator<ServiceProvider>() {
+      @Override
+      public int compare(ServiceProvider f1, ServiceProvider f2) {
+        return f1.priority() - f2.priority();
+      }
+    }));
+    return Collections.unmodifiableList(list);
+  }
+
+  /**
+   * Returns true if the {@link ClassLoader} is for android.
+   */
+  static boolean isAndroid(ClassLoader cl) {
+    try {
+      // Specify a class loader instead of null because we may be running under Robolectric
+      Class.forName("android.app.Application", /*initialize=*/ false, cl);
+      return true;
+    } catch (Exception e) {
+      // If Application isn't loaded, it might as well not be Android.
+      return false;
+    }
+  }
+
+  /**
+   * Loads service providers for the {@link ServiceProvider} service using {@link ServiceLoader}.
+   */
+  @VisibleForTesting
+  public static <T> Iterable<T> getCandidatesViaServiceLoader(Class<T> klass, ClassLoader cl) {
+    Iterable<T> i = ServiceLoader.load(klass, cl);
+    // Attempt to load using the context class loader and ServiceLoader.
+    // This allows frameworks like http://aries.apache.org/modules/spi-fly.html to plug in.
+    if (!i.iterator().hasNext()) {
+      i = ServiceLoader.load(klass);
+    }
+    return i;
+  }
+
+  /**
+   * Load providers from a hard-coded list. This avoids using getResource(), which has performance
+   * problems on Android (see https://github.com/grpc/grpc-java/issues/2037).
+   */
+  @VisibleForTesting
+  static <T> Iterable<T> getCandidatesViaHardCoded(
+      Class<T> klass, List<String> classNames, ClassLoader cl) {
+    // Class.forName(String) is used to remove the need for ProGuard configuration. Note that
+    // ProGuard does not detect usages of Class.forName(String, boolean, ClassLoader):
+    // https://sourceforge.net/p/proguard/bugs/418/
+    List<T> list = new ArrayList<T>();
+    for (String name : classNames) {
+      try {
+        list.add(create(klass, cl.loadClass(name)));
```

Comment:
This doesn't work with ProGuard. It needs to be the exact literals like `Class.forName("io.grpc.okhttp.OkHttpChannelProvider")` we're using today, otherwise ProGuard won't detect the string and won't keep needed classes and won't rename properly. This is also why the hard-coded code doesn't use a `ClassLoader` (which is what that comment above is about).

Reply:
Changed back to old behavior.